### PR TITLE
Prio

### PR DESCRIPTION
**Pull request by wadaniel**
Hi, here is the prioritized replay memory implementation,
I integrated it in NAF and it improved the learning. 
Note that each learner needs to call two additional functions I the train method:
updateExperienceReplayPriority and calculateImportanceWeights

once we approved this PR, I will do the integration for the other learners.
feedback & discussions welcome

### CODE REVIEW

**Review by SergioMartin86 — APPROVED** [review]

File: source/modules/solver/agent/agent._cpp [hunk 1]
```diff
@@ -194,11 +187,50 @@ void Agent::processExperience(knlohmann::json &experience)
...
+  sort(experienceRanking.begin(), experienceRanking.end());
+
+  // Calculate rank based probabilities
+  float prioritySum = 0.0;
+  for (size_t i = 1; i < _experienceReplayStates.size() + 1; ++i)
+    prioritySum += pow((float)i, _priorityAnnealingRate);
+
+  // Update probabilities
+  for (size_t i = 0; i < _experienceReplayStates.size(); ++i)
+    _experienceReplayProbabilities[i] = pow((float)experienceRanking[i].second, _priorityAnnealingRate) / (prioritySum);
```
Comment (webepasc) [inline source/modules/solver/agent/agent._cpp:211]: no std:: needed here?

File: source/modules/solver/agent/agent._cpp [hunk 2]
```diff
@@ -194,11 +187,50 @@ void Agent::processExperience(knlohmann::json &experience)
...
+}
+
+std::vector<float> Agent::calculateImportanceWeights(std::vector<size_t> miniBatch)
+{
+  auto importanceWeights = std::vector<float>(miniBatch.size(), 1.0);
+
+  if (_miniBatchStrategy == "Prioritized")
+  {
+    for (size_t i = 0; i < miniBatch.size(); ++i)
+      importanceWeights[i] = pow((float)_experienceReplayStates.size() * _experienceReplayProbabilities[miniBatch[i]], -_importanceWeightAnnealingRate);
```
Comment (webepasc) [inline source/modules/solver/agent/agent._cpp:228]: also std::

File: source/modules/solver/agent/agent._cpp [hunk 3]
```diff
@@ -218,10 +250,62 @@ std::vector<size_t> Agent::generateMiniBatch(size_t size)
...
+  else if (_miniBatchStrategy == "Prioritized")
+  {
+    // Allocating storage for uniform random numbers
+    auto pvalues = std::vector<float>(size);
+
+    for (size_t i = 0; i < size; i++)
+      pvalues[i] = _uniformGenerator->getRandomNumber();
+
+    // Sort selections ascending for traversal
+    sort(pvalues.begin(), pvalues.end());
```
Comment (webepasc) [inline source/modules/solver/agent/agent._cpp:262]: std::

File: source/modules/solver/agent/agent._cpp [hunk 4]
```diff
@@ -218,10 +250,62 @@ std::vector<size_t> Agent::generateMiniBatch(size_t size)
...
+    while (batchIdx < size && expId < _experienceReplayStates.size())
+    {
+      if (cumulativeProbability > pvalues[batchIdx])
+        miniBatch[batchIdx++] = expId;
+      else
+        cumulativeProbability += _experienceReplayProbabilities[++expId];
+    }
+
+    // Uniform fill if miniBatch not full due to numerical precision during probability accumulation (cumulative probability < 1)
+    while (batchIdx < size) miniBatch[batchIdx++] = floor(_uniformGenerator->getRandomNumber() * (float)(_experienceReplayStates.size() - 1));
```
Comment (webepasc) [inline source/modules/solver/agent/agent._cpp:279]: std::

File: source/modules/solver/agent/agent.config [hunk 5]
```diff
@@ -130,6 +140,7 @@
    "Type": "std::string",
    "Options": [
       { "Value": "Uniform", "Description": "Selects experiences from the replay memory with a randomly." },
+      { "Value": "Prioritized", "Description": "Prioritizes experiences according to their rank." },
```
Comment (webepasc) [inline source/modules/solver/agent/agent.config:143]: should we add references here? Or where else?

**Review by webepasc — COMMENTED** [review]
Maybe you want to go over my comments and adjust

**Review by SergioMartin86 — APPROVED** [review]

**Pull request merged**